src/webapp/components/form/FormSectionsState.ts

```diff
@@ -154,3 +156,48 @@ function validateSection(
         }
     });
 }
+
+// RULES:
+
+export function toggleSectionVisibilityByFieldValue(
+    section: FormSectionState,
+    fieldValue: FormFieldState["value"],
+    rule: Rule
+): FormSectionState {
+    if (rule.sectionIds.includes(section.id)) {
+        const subsections = section.subsections?.map(subsection => {
+            return toggleSectionVisibilityByFieldValue(subsection, fieldValue, rule);
+        });
+        return section.subsections
+            ? {
+                  ...section,
+                  subsections: subsections,
+                  isVisible: fieldValue === rule.fieldValue,
+                  fields:
+                      fieldValue === rule.fieldValue
+                          ? section.fields.map(field => ({
+                                ...field,
+                                isVisible: true,
+                            }))
+                          : hideFieldsAndSetToEmpty(section.fields),
+              }
+            : {
+                  ...section,
+                  isVisible: fieldValue === rule.fieldValue,
+                  fields:
+                      fieldValue === rule.fieldValue
+                          ? section.fields.map(field => ({
+                                ...field,
+                                isVisible: true,
+                            }))
+                          : hideFieldsAndSetToEmpty(section.fields),
+              };
```

Comment:
the two expressions are very similar, you can create a local variable before that captures the common fields.